lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs

```diff
@@ -0,0 +1,210 @@
+                .FirstOrDefault();
+
+            AssertCancelEvent(observer, false);
+        }
+
+        [Fact]
+        public void JobLifeCyclemanger_DetectorNull_DoesNotSendEvent()
+        {
+            var observer = JobLifeCycleMangerEventTest(
+                detector: null)
+                .FirstOrDefault();
+
+            AssertCancelEvent(observer, false);
+        }
+
+        [Fact]
+        public void JobLifeCyclemanger_NoObservers_DoesNotCheckForSignal()
+        {
+            int isCancelledCheckCounter = 0;
+
+            var observer = JobLifeCycleMangerEventTest(
+                detector: new DetectorStubStatic(true, "cancelled", testAction: () => { isCancelledCheckCounter++; }),
+                subscribeObserver: false,
+                signalCheckPeriodSec: 1,
+                waitForEventPeriodSec: 6)
+                .FirstOrDefault();
+
+            Assert.True(isCancelledCheckCounter == 0, "Expected no checks for cancellation if there are no subscribers. Actual check counter: " + isCancelledCheckCounter);
+            AssertCancelEvent(observer, false);
+        }
+
+        private IEnumerable<TestObserver> JobLifeCycleMangerEventTest(
+            IJobCancelledDetector detector,
+            bool subscribeObserver = true,
+            int observerCount = 1,
+            int signalCheckPeriodSec = 1,
+            int waitForEventPeriodSec = 2)
+        {
+            var manager = Activator.CreateInstance(
+                typeof(JobLifeCycleManager),
+                BindingFlags.NonPublic | BindingFlags.Instance,
+                null,
+                new object[] { detector, signalCheckPeriodSec },
+                null,
+                null) as JobLifeCycleManager;
+
+            var observers = Enumerable.Range(1, observerCount)
+                .Select(_ => subscribeObserver ? new TestObserver(manager) : new TestObserver(null))
+                .ToList();
+
+            manager.OnNext(NewStartedEvent());
+
+            Thread.Sleep(waitForEventPeriodSec * 1000);
+
+            return observers;
+        }
+
+        private void AssertCancelEvent(TestObserver observer, bool expectedEvent, string expectedMessage = null)
+        {
+            if (expectedEvent)
+            {
+                Assert.NotNull(observer.LastEvent);
+                Assert.Same(expectedMessage, observer.LastEvent.Message);
+            }
+            else
+            {
+                Assert.Null(observer.LastEvent);
+            }
+        }
+
+        private IDriverStarted NewStartedEvent()
+        {
+            // event is not really used by the driver, so can use null here
+            return null;
+        }
+
+        /// <summary>
+        /// Test helper class to provide predefined cancel signal for testing
+        /// </summary>
+        private class DetectorStubStatic : IJobCancelledDetector 
+        {
+            private bool IsCancelledResponse { get; set; }
```

Comment:
Private properties are less commonly used than private variables, unless you want to wrap some logic inside the get/set methods. 
Property access will be (fractionally) slower as it will call the get/set. 
So if you only have simple get/set, my opinion is private variable is better.